### PR TITLE
[Complete in first approximation] iOS and general touchscreen support, pass 4

### PR DESCRIPTION
**Pull request by singalen**
Applied review notes from PR #1742, #1748. Reincarnation of PR #1752 rebased onto current master.

It is STILL not in a shape I'd like to see it, but it's playable on a touchscreen.

Known problems: https://github.com/singalen/wesnoth/labels/bug

### CODE REVIEW

**Comment by CelticMinstrel** [comment]
I see maybe two bugs on that list that would be relevant for a Windows tablet or a PC (Windows, Mac, or Linux) with a touchscreen. The rest are very specific to iOS and have nothing to do with whether this should be merged to master.

Personally, I think this _should_ be merged, but I'm a little worried about accepting such a huge pull request when we're supposed to be moving towards release. We're not on feature freeze yet though, so perhaps it would be acceptable. Has anyone tested this on touch devices other than iOS?

**Comment by gfgtdf** [comment]
some things i noticed:
1) you use gui2::add_timer in controller_base, since controller base is not a gui2 dialog i'd be surprised if that worked.
2) You use the `#if defined(__APPLE__) && defined(__MACH__) && defined(__ENVIRONMENT_IPHONE_OS_VERSION_MIN_REQUIRED__)
#define __IPHONEOS__
#endif` You shoudl put it in some header, maybe even global.hpp

**Comment by singalen** [comment]
Thanks @gfgtdf.

1. It works once you handle `TIMER_EVENT` (I added a handler). I use it instead of a bare SDL timer to get the callback on the main thread.
2. I totally agree. Sadly, `global.hpp` is the only place that I've seen fit for it. I didn't like the idea of starting a global scrapheap of code that doesn't fit anywhere else, or to create a new header just for that. But if you say it's OK I will :)

**Comment by singalen** [comment]
@CelticMinstrel,

The main point of the PR (for me) is to keep iOS up with master. Otherwise rebasing onto master will yield more and more merges for me (and already does!).

Another point is to test touch controls on PC.
@Vultraz has ran an early version on his touchscreen laptop, probably he can build this one when idle and test it.
BTW one can add `#define MOUSE_TOUCH_EMULATION` right into three `cpp` files (in order not to rebuild everything) and more or less test touch interface with RMB.

I'm playing it on iOS frequently, plus now we have a group of TestFlight beta testers.

File: src/controller_base.cpp [hunk 1]
```diff
@@ -38,11 +41,41 @@ controller_base::controller_base(const config& game_config)
...
 
 controller_base::~controller_base()
 {
+	if(long_touch_timer_ != 0) {
+		gui2::remove_timer(long_touch_timer_);
+        long_touch_timer_ = 0;
+    }
+}
+
+void controller_base::long_touch_callback(int x, int y)
```
Comment (Vultraz) [inline src/controller_base.cpp]: Indentation is all messed up in this function. Please convert it to tabs.

Comment (Vultraz) [inline src/controller_base.cpp]: Indentation is all messed up in this function. Please convert it to tabs.

Comment (singalen) [inline src/controller_base.cpp]: Fixed.

File: src/controller_base.cpp [hunk 2]
```diff
@@ -17,17 +17,20 @@
...
 #include "mouse_handler_base.hpp"
 #include "preferences/game.hpp"
 #include "scripting/plugins/context.hpp"
 #include "show_dialog.hpp" //gui::in_dialog
 #include "soundsource.hpp"
+#include "gui/core/timer.hpp"
+
 static lg::log_domain log_display("display");
 #define ERR_DP LOG_STREAM(err, log_display)
+#define LONG_TOUCH_DURATION_MS 800
```
Comment (gfgtdf) [inline src/controller_base.cpp]: we usually prefer (static) const for such constants instead of macros.

**Comment by Vultraz** [comment]
Pushing back to 1.13.10.

File: src/controller_base.cpp (hunk 2 above)
Comment (singalen) [reply src/controller_base.cpp]: Replaced with `static const int`.

**Comment by Vultraz** [comment]
Closing this in favor of some new PRs @singalen intends to open.

**Pull request closed without merging**